src/nodes/kvs/store.rs

```diff
@@ -0,0 +1,184 @@
+use crate::{
+    lattice::{
+        causal::SingleKeyCausalLattice,
+        last_writer_wins::{Timestamp, TimestampValuePair},
+        CounterLattice, LastWriterWinsLattice, Lattice, MapLattice, MaxLattice, SetLattice,
+    },
+    messages::{
+        request::{ClientKeyOperation, InnerKeyOperation},
+        response::ClientResponseValue,
+    },
+    store::LatticeValue,
+    AnnaError, Key,
+};
+
+use super::KvsNode;
+
+impl KvsNode {
+    pub(crate) fn key_operation_handler(
+        &mut self,
+        operation: ClientKeyOperation,
+        timestamp: chrono::DateTime<chrono::Utc>,
+    ) -> Result<Option<ClientResponseValue>, AnnaError> {
+        use std::collections::hash_map::Entry;
+
+        match operation {
```

Comment:
Would it be possible to always go through `inner_key_operation_handler`? My idea in https://github.com/essa-project/anna-rs/pull/4#issuecomment-1375514380 was that we could translate the `ClientKeyOperation` to an `InnerKeyOperation` that uses the normal merge logic. This way, we could be sure that we don't violate any invariants of the lattice types (e.g. that a value is monotonically increasing).

For example, instead of doing an in-place modification of `SingleCausal` lattice values, we could construct a new `SingleKeyCausalLattice` and then invoke `inner_key_operation_handler`. This way, we also reduce code duplication.

What do you think?